dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,208 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_JOIN_TENSOR_JOIN_CPU_H_
+#define DALI_KERNELS_COMMON_JOIN_TENSOR_JOIN_CPU_H_
+
+#include <vector>
+#include "dali/kernels/kernel.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/core/format.h"
+
+namespace dali {
+namespace kernels {
+namespace impl {
+
+template<typename T>
+void
+ConcatenateTensors(TensorView<StorageCPU, T> output,
+                   span<const TensorView<StorageCPU, const T>> inputs,
+                   int axis) {
+  SmallVector<int64_t, 8> copy_sizes;
+  for (int t = 0; t < inputs.size(); t++) {
+    copy_sizes[t] = volume(inputs[t].shape.begin() + axis, inputs[t].shape.end());
+  }
+  auto nouter = volume(output.shape.begin(), output.shape.begin() + axis);
+  auto *out = output.data;
+  for (ptrdiff_t outer = 0; outer < nouter; outer++) {
+    for (int t = 0; t < inputs.size(); t++) {
+      auto *src = inputs[t].data + outer * copy_sizes[t];
+      for (ptrdiff_t inner = 0; inner < copy_sizes[t]; inner++) {
+        *out++ = src[inner];
+      }
+    }
+  }
+}
+
+
+template<bool new_axis = true>
+TensorShape<> DetermineShape(span<const TensorShape<>> in_shapes, int axis) {
+  TensorShape<> ret;
+  auto &insh = in_shapes[0];
+  ret.resize(insh.size() + 1);
+  for (int i = 0; i < axis; i++) {
+    ret[i] = insh[i];
+  }
+  ret[axis] = in_shapes.size();
+  for (int i = axis + 1; i < ret.size(); i++) {
+    ret[i] = insh[i - 1];
+  }
+  return ret;
+}
+
+
+template<>
+TensorShape<> DetermineShape<false>(span<const TensorShape<>> in_shapes, int axis) {
+  TensorShape<> ret = in_shapes[0];
+  ret[axis] = 0;
+  for (const auto &sh : in_shapes) {
+    ret[axis] += sh[axis];
+  }
+  return ret;
+}
+
+
+/**
+ * Joins multiple input tensors into one output tensor, along given axis.
+ *
+ * The kernel works in 2 modes: STACK and CONCAT. In CONCAT mode, kernel creates a new tensor,
+ * with joined values along given dimension, e.g.
+ *
+ * arr0 = [[1, 2, 4, 2], [1, 1, 7, 6], [6, 8, 8, 4]]
+ * shape = (3, 4)
+ *
+ * arr1 = [[3, 8, 8, 6], [8, 1, 5, 7], [6, 2, 7, 5]]
+ * shape = (3, 4)
+ *
+ * concatenate([arr0, arr1], axis=1) -> [[1, 2, 4, 2, 3, 8, 8, 6],
+ *                                       [1, 1, 7, 6, 8, 1, 5, 7],
+ *                                       [6, 8, 8, 4, 6, 2, 7, 5]]
+ * shape = (3, 8)
+ *
+ * Stacking, on the other hand, creates new tensor with added dimension, where `axis` points.
+ *
+ * stack([arr0, arr1], axis=1) -> [[[1, 2, 4, 2],
+ *                                  [3, 8, 8, 6]],
+ *
+ *                                 [[1, 1, 7, 6],
+ *                                  [8, 1, 5, 7]],
+ *
+ *                                 [[6, 8, 8, 4],
+ *                                  [6, 2, 7, 5]]]
+ * shape = (3, 2, 4)
+ *
+ * @tparam new_axis if true, STACK mode is applied.
+ */
+template<typename T, bool new_axis = true, int dims = -1>
```

Comment:
Maybe `new_axis` should be named as `stack_mode`?

Reply:
`new_axis` is what I agreed on with @mzient